### PR TITLE
Links, simple manifest and redis/postgres target jobs

### PR DESCRIPTION
**Pull request by drnic**
This PR adds bosh links to jobs, a new manifest that assumes cloud-config, and includes two new jobs for simple targeting/backing up of a postgres or redis. This release now assumes using `bosh2`. The new manifest generates passwords/certs.

To test/demo:

```
git clone https://github.com/starkandwayne/shield-boshrelease -b links-and-cloud-config
cd shield-boshrelease
git submodule update --init
bosh2 create-release
bosh2 upload-release --rebase
bosh2 deploy manifests/shield.yml -d shield --vars-store=tmp/shield-creds.yml
```

* Run `bosh2 instances` to get the IP (e.g. `10.244.0.2`).
* Run `bosh2 int tmp/shield-creds.yml --path /shield-daemon-password` to get basic auth password (user is `admin`)

To add `s3` as a store:

```
bosh2 deploy manifests/shield.yml -d shield --vars-store=tmp/shield-creds.yml \
  -o manifests/operators/store-amazon-s3.yml \
  -v s3-access-key=x
  -v s3-secret-key=x
  -v s3-bucket=x
  -v s3-bucket-store-prefix=x
```

To test postgres with backups:

```
bosh2 deploy manifests/target-examples/postgres.yml --vars-store=tmp/postgres-creds.yml -d shield-target-postgres
```

To test redis with backups:

```
bosh2 deploy manifests/target-examples/redis.yml --vars-store=tmp/redis-creds.yml -d shield-target-redis
```

Targets will look like:

screen shot 2017-05-08 at 10 35 09 am

These manifests have also been tested with credhub.

### Todo
- [x] merge nginx into shield-daemon so to stop clients of daemon connecting via :8080

### CODE REVIEW

**Comment by drnic** [comment]
This PR has two of https://github.com/starkandwayne/shield-boshrelease/pull/75 fixes included. I'd like https://github.com/starkandwayne/shield-boshrelease/pull/75/commits/5f7bab2719dd9267fc6e6d13f740495c0b0e3429 as well.

File: jobs/shield-agent/templates/bin/ctl [hunk 1]
```diff
@@ -14,11 +14,19 @@ case $1 in
     echo "Attempting to set up SSH public keys"
     AUTHKEYS=/var/vcap/jobs/shield-agent/config/authorized_keys
     rm -f $AUTHKEYS
-    <% if_p("shield.agent.autoprovision") do |endpoint| %>curl -Lks "<%= endpoint %>/v1/meta/pubkey" -m 20 >> $AUTHKEYS || true<% end %>
-    <% if_p("shield.agent.daemon_public_key") do |key| %>cat >>$AUTHKEYS <<EOF
+    <%
+      if_p("autoprovision") do |endpoint|
+        if endpoint.to_s == "true"
+          endpoint = "http://#{link('shield-daemon').instances.first.address}:8080"
```
Comment (frodenas) [inline jobs/shield-agent/templates/bin/ctl]: we should use the `https` endpoint (secure by default). Also, the link should be optional (`if_link`), see my previous comment.

File: jobs/shield-daemon/spec [hunk 2]
```diff
@@ -9,68 +9,83 @@ templates:
...
+  properties:
+  - domain
+  - port
+  - http_port
+  - auth.api_keys
+
+consumes:
+- name: shield-db
+  type: sql
+- name: sessions-db
```
Comment (frodenas) [inline jobs/shield-daemon/spec]: This should be optional, it's only needed if oauth is used, otherwise, we don't need the sessions db.

**Comment by frodenas** [comment]
@drnic Beside the above comments, I don't understand why you want to add a `shield-target-postgres` and `shield-target-redis` jobs. You can do exactly the same by:

1) collocating the `shield-agent` job (and the backup/restore will take place at the same vm where postgres or redis resides)
2) converting this to a generic errand (that just registers the job)

Another problem that I see by using these jobs is that they don't include the `shield-agent` package, and they're registering the job using the vm ip address. So if you don't add the `shield-agent`, the backup will fail because the `shield-daemon` will try to contact that ip address and there's no `shield-agent` to respond.

**Comment by drnic** [comment]
@frodenas (and thanks @geofffranks for chat) ok I've gotten rid of `shield-target-{postgres,redis}` - I believe I got lost in the assumed complexity of links; ultimately the two jobs had to connect to postgres/redis without links anyway. 

See postgres.yml + redis.yml examples now https://github.com/starkandwayne/shield-boshrelease/tree/links-and-cloud-config/manifests/target-examples 

They are using the `shield-agent`'s `targets` and `jobs` properties.

File: jobs/shield-agent/templates/bin/ctl (hunk 1 above)
Comment (drnic) [inline jobs/shield-agent/templates/bin/ctl]: Agreed. @geofffranks agreed to me merging nginx into shield-daemon job, so that one nice link will communicate the public https information; and we can close off :8080 again.

File: jobs/nginx/monit [hunk 3]
```diff
@@ -1,5 +1,5 @@
 check process nginx
   with pidfile /var/vcap/sys/run/nginx/nginx.pid
```
Comment (drnic) [inline jobs/nginx/monit]: Thanks. Accidental search/replace error.

File: jobs/nginx/spec [hunk 4]
```diff
@@ -4,49 +4,50 @@ packages:
...
   config/ssl_key:    config/ssl_key
   config/ssl_crt:    config/ssl_crt
 
+consumes:
+- name: shield-daemon
+  type: shield-daemon
+
 properties:
-  shield.daemon.port:
+  port:
```
Comment (drnic) [inline jobs/nginx/spec]: I'd like to get it to just `port` as 443, assume https, and stop providing `http_port`.

Or if https is disabled, then `port` would be explicitly `80`.

@geofffranks @frodenas ideas/preference to moving to single `https`/`port: 443` property?

File: jobs/shield-agent/templates/bin/ctl (hunk 1 above)
Comment (drnic) [inline jobs/shield-agent/templates/bin/ctl]: I'm going to move towards mandatory https for shield-daemon, since it's as easy to setup https as not to with the new manifest + bosh2

**Comment by drnic** [comment]
I'm concerned that I've not got ssl correctly configured. Should the following curl command work if I got it right; or is my sanity test wrong?

```
./bin/shield-cacert > tmp/ssh/cacert
curl -u admin:$(./bin/shield-password) https://10.244.0.2 --cacert tmp/ssh/cacert -v
```
The output is:

```
* Rebuilt URL to: https://10.244.0.2/
*   Trying 10.244.0.2...
* Connected to 10.244.0.2 (10.244.0.2) port 443 (#0)
* WARNING: using IP address, SNI is being disabled by the OS.
* SSL: certificate verification failed (result: 5)
* Closing connection 0
curl: (51) SSL: certificate verification failed (result: 5)
```

File: jobs/shield-agent/spec [hunk 5]
```diff
@@ -12,45 +12,51 @@ templates:
   bin/post-start:           bin/post-start
   bin/emergency-recovery:   bin/emergency-recovery
   config/shield-agent.conf: config/shield-agent.conf
-  data/properties.sh:       data/properties.sh
   helpers/ctl_setup.sh:     helpers/ctl_setup.sh
   helpers/ctl_utils.sh:     helpers/ctl_utils.sh
 
+consumes:
+- name: shield-daemon
```
Comment (drnic) [inline jobs/shield-agent/spec]: Now optional, thanks.

File: jobs/shield-agent/templates/bin/ctl (hunk 1 above)
Comment (drnic) [inline jobs/shield-agent/templates/bin/ctl]: Fixed to now be:

```
      domain = p("domain", link('shield-daemon').instances.first.address)
      endpoint = "https://#{domain}"
```

File: jobs/shield-daemon/spec (hunk 2 above)
Comment (drnic) [inline jobs/shield-daemon/spec]: Fixed, thanks. And the rest of the templates support explicit `p('database.host')` declaration of database.

**Comment by drnic** [comment]
@frodenas @geofffranks I believe the PR is updated based on feedback from both of you. Additional round of reviews/feedback is warmly welcome!

**Comment by drnic** [comment]
screen shot 2017-05-09 at 5 32 02 pm

I'm now getting this in Chrome. I don't get it in Safari. Ideas?

**Comment by drnic** [comment]
screen shot 2017-05-09 at 5 33 17 pm

On Safari

File: jobs/mariadb/monit [hunk 6]
```diff
@@ -1,5 +1,5 @@
 check process mariadb
   with pidfile /var/vcap/sys/run/mariadb/mariadb.pid
-  start program "/var/vcap/jobs/mariadb/bin/mariadb_ctl start" timeout 780 seconds
-  stop program "/var/vcap/jobs/mariadb/bin/mariadb_ctl stop"
+  start program "/var/vcap/jobs/mariadb/bin/ctl start" timeout 780 seconds
+  stop program "/var/vcap/jobs/mariadb/bin/ctl stop"
```
Comment (jhunt) [inline jobs/mariadb/monit]: Is this change required for supporting links?

File: jobs/postgres/monit [hunk 7]
```diff
@@ -1,6 +1,6 @@
 check process postgres
   with pidfile /var/vcap/sys/run/postgres/postgres.pid
-  start program "/var/vcap/jobs/postgres/bin/postgres_ctl start" timeout 780 seconds
-  stop program "/var/vcap/jobs/postgres/bin/postgres_ctl stop"
+  start program "/var/vcap/jobs/postgres/bin/ctl start" timeout 780 seconds
+  stop program "/var/vcap/jobs/postgres/bin/ctl stop"
```
Comment (jhunt) [inline jobs/postgres/monit]: Is this change required for supporting links?

File: jobs/postgres/spec [hunk 8]
```diff
@@ -5,26 +5,40 @@ packages:
...
+
 properties:
-  databases.port:
+  port:
     description: Port to listen on
+    default: 5432
+
+  db_scheme:
+    description: Link property to declare it as postgres://
+    default: postgres
```
Comment (jhunt) [inline jobs/postgres/spec]: These are backwards-compatibility breaking changes.  Are they strictly necessary to support links, or would these changes be better served in a separate PR that can be merged with a major version bump?

File: jobs/shield-agent/templates/bin/post-start [hunk 9]
```diff
@@ -1,47 +1,56 @@
...
+export SHIELD_API_TOKEN=<%= api_key %>
+<% if p("skip_ssl_verify") -%>
+export SHIELD_SKIP_SSL_VERIFY=true
+<% end -%>
 
 # Create a specific  config file for shield-agent stuff, to avoid race
-# conditions in case we are colocated with shield-daemon. %>
-shield -c ${SHIELD_CONFIG} create backend default <%= server %>
+# conditions in case we are colocated with shield-daemon.
+shield -c ${SHIELD_CONFIG} create-backend default <%= endpoint %>
```
Comment (jhunt) [inline jobs/shield-agent/templates/bin/post-start]: This change also doesn't seem to be related to links...

File: jobs/shield-daemon/templates/config/shieldd.conf [hunk 10]
```diff
@@ -1,40 +1,75 @@
...
+<% end -%>
+<% if_p("auth.oauth.provider") do |provider| -%>
   oauth:
     provider: <%= provider %>
-    key: <%= p("shield.daemon.auth.oauth.key") %>
-    secret: <%= p("shield.daemon.auth.oauth.secret") %>
-    base_url: https://<%= p("shield.daemon.domain") %>:<%= p("shield.daemon.port") %>
+    key: <%= p("auth.oauth.key") %>
+    secret: <%= p("auth.oauth.secret") %>
+    base_url: https://<%= p("domain", link("shield-daemon").instances.first.address) %>:<%= p("port") %>
```
Comment (jhunt) [inline jobs/shield-daemon/templates/config/shieldd.conf]: Does this work if the SHIELD instance is load balanaced / proxy-forwarded by an external piece of equipment with a different name / IP?  IIRC, that was the purpose of the `shield.daemone.domain` setting.

File: jobs/shield-daemon/templates/nginx/config/nginx.conf [hunk 11]
```diff
@@ -37,46 +37,32 @@ http {
   gzip_disable       "MSIE [1-6]\.(?!.*SV1)";
 
   upstream shield {
-    server localhost:8080;
+    server 127.0.0.1:8080;
```
Comment (jhunt) [inline jobs/shield-daemon/templates/nginx/config/nginx.conf]: Was there a reason for this change from name to IP?

File: jobs/shield-agent/templates/bin/post-start (hunk 9 above)
Comment (jhunt) [inline jobs/shield-agent/templates/bin/post-start]: Ah, I see now.  This was from picking up #75

**Comment by drnic** [comment]
screen shot 2017-05-10 at 11 03 59 am

Added `testflight-links` (to bosh-lite49 with cloud-config)

File: jobs/postgres/spec (hunk 8 above)
Comment (drnic) [inline jobs/postgres/spec]: Yeah this one I think is explicitly for links - to expose the type of the db for consumers. Not really for overriding; just for passing out to the consumers of the link.

File: jobs/shield-daemon/templates/nginx/config/nginx.conf (hunk 11 above)
Comment (drnic) [inline jobs/shield-daemon/templates/nginx/config/nginx.conf]: it was changed to 0.0.0.0 in the time before I merged nginx into daemon job; then changed "back" to the valid as I remembered it. I will revert to `localhost`.

File: jobs/shield-daemon/templates/config/shieldd.conf (hunk 10 above)
Comment (drnic) [inline jobs/shield-daemon/templates/config/shieldd.conf]: Yeah, hopefully it works like that. I've only tested the link usage. I kept `domain` as I assumed ppl were using it for external routes/https. I will add an operations file to do route registrar and set `domain`

**Comment by drnic** [comment]
Rebased against v6.8.0

**Comment by jhunt** [comment]
With the 6.x branch of the BOSH release safely tucked away for the production users, I have no reservations merging this, if it's ready.  @drnic - anything left to do / test / fix on this?

**Comment by drnic** [comment]
I'd like someone to check the setup of SSL certs. On my own deployments, Chrome is unhappy.

**Comment by jhunt** [comment]
I'm getting this when I deploy the shield.yml to my BOSH-lite:

```
2017-05-15 14:44:49.368763104 +0000 UTC shieldd: ERROR: worker 5 unable to read user key /var/vcap/jobs/shield-daemon/shared/id_rsa: ssh: no key found; bailing out.
```

**Comment by drnic** [comment]
@jhunt when you did your deploy, did you use `--vars-store tmp/creds.yml` or a bosh with credhub? At a guess, this is an error because your `/var/vcap/jobs/shield-daemon/shared/id_rsa` file contains nonsense.

At a guess, I bet that file contains the string `((shield-daemon-sshkey.private_key))`.

Unfortunately `bosh2 deploy` & director do not error if any variables are not provided before deployment is commenced.

**Comment by drnic** [comment]
Ok, confirmed that SSL certs are working. Needed a way to add public IP into the cert:

```
bosh2 deploy manifests/shield.yml \
  --vars-store tmp/creds.yml \
  -o manifests/operators/tls-alternative-name.yml -v tls-alternative-name=10.58.111.50
```

And then the following works as expected:

```
bosh2 int tmp/creds.yml --path /shield-tls/ca > tmp/ssh/cacert-lite50
curl https://10.58.111.50:8443/ -u admin:admin --cacert tmp/ssh/cacert-lite50
```

**Comment by drnic** [comment]
@jhunt I feel like this PR is good now.

I've got a draft `doc/upgrade-to-cloud-config.md` doc; but it focuses on the only known starting point - `templates/make_manifest warden` which is not a production upgrade scenario.

So perhaps once we have a shield v7.0.0.rc-1 or similar, we can look at some production upgrades and see what notes would be helpful?

**Comment by jhunt** [comment]
I did not have a credentials storage (file or credhub).

Sure would be nice if BOSH would catch these errors.  Until then, how do you feel about a change to check for errors in templates?

```
18:20:57 | Error: Unable to render instance groups for deployment. Errors are:
   - Unable to render jobs for instance group 'shield'. Errors are:
     - Unable to render templates for job 'shield-daemon'. Errors are:
       - Error filling in template 'id_rsa' (line 6: ssh_private_key '((shield-daemon-sshkey.private_key))' does not look like an RSA private key)
```

**Comment by jhunt** [comment]
Is there a way to generate the credentials (without using credhub) in-repo?
